uwblueprint/starter-code-v2
# Create Python FileStorageService

## Notion ticket link

[Create Python FileStorageService](https://www.notion.so/uwblueprintexecs/Task-Board-db95cd7b93f245f78ee85e3a8a6a316d?p=92d1b47901da4b128aff4285bd5da57b)



## Implementation description
* Added a file storage service for blobs with create, update, get, and delete functions
* Create endpoint only creates if file name doesn't already exist, if it does we throw an exception
* Update endpoint only updates if file name exists, if not we throw an exception


## Steps to test
1. Visit the Firebase console for this project and navigate to Storage
<img width="240" alt="image" src="https://user-images.githubusercontent.com/37782734/123180950-5699d780-d45a-11eb-8f9f-a4a5aab62744.png">
2. Add a file you want to test with to your /app directory

3. Add the DEFAULT_BUCKET env value. You may also want to change `erase_db_and_sync` to be False in models/\_\_init\_\_.py so that the terminal isn't bloated when running

4. Add the following lines of code to the root \_\_init\_\_.py

```
from firebase_admin import credentials
from flask import current_app
from .services.implementations.file_storage_service import FileStorageService
```

After `rest.init_app(app)` add the following lines of code
```
    file_storage_service = FileStorageService(current_app.logger)
    script_dir = os.path.dirname(__file__)
    rel_path = "your_file.ext"
    abs_file_path = os.path.join(script_dir, rel_path)

    # Get File
    # print(file_storage_service.get_file("your_file.ext"))

    # Create File
    # with open(abs_file_path, 'rb') as file:
    #    print(file_storage_service.create_file('your_file.ext', file))

    # Update File
    # with open(abs_file_path, 'rb') as file:
    #    print(file_storage_service.update_file('your_file.ext', file))
    
    # Delete File
    # file_storage_service.delete_file("your_file.ext")
```
5. Start the container with docker-compose up --build
6. Uncomment Create File and verify that your file is created in the Firebase dashboard. Also verify that you get the signed url to the file. Run it again and verify that you get an exception. 
7. Comment Create File and uncomment Get File. Verify that the signed url to the file is printed
8. Comment Get File and uncomment Update File. Verify that your file is updated in the Firebase dashboard. Also verify that you get the signed url to the file. Change the string argument to another file, and verify that you get an exception.
9. Comment Update File and uncomment Delete File. Verify that your file is deleted in the Firebase dashboard. Change the string argument to another file and verify that you get an exception
10. Comment Delete File and uncomment Get File. Verify that None is printed

## What should reviewers focus on?
* Are we handling exceptions properly?
* Does the behaviour of create and update make sense? An alternative is that we could have a single function that always either creates or replaces a file without any checks


## Checklist
